qemu/tests/numa_opts.py

```diff
@@ -1,49 +1,115 @@
+        """
+
+        numa_info_guest = NumaInfo(session=session)
+
+        numa_guest = []
+        nodes_guest = numa_info_guest.online_nodes
+        for node in nodes_guest:
+            node_size = numa_info_guest.online_nodes_meminfo[node]['MemTotal']
+            node_size = float(normalize_data_size('%s KB' % node_size))
+            node_cpus = numa_info_guest.online_nodes_cpus[node]
+            node_cpus = set([int(v) for v in node_cpus.split()])
+            numa_guest.append((node_size, node_cpus))
+
+        # It is a known WONTFIX issue for x86, node info of node0 and node1 is
+        # opposite in guest os when vm have 2 nodes
+        if (vm_arch in ("x86_64", "i686") and len(numa_guest) == 2):
+            numa_guest.reverse()
+        return numa_guest
 
     vm = env.get_vm(params["main_vm"])
+    os_type = params["os_type"]
+    vm_arch = params["vm_arch_name"]
+    session = vm.wait_for_login()
 
-    numa = vm.monitors[0].info_numa()
-    dbg("info numa reply: %r", numa)
+    error_context.context("starting numa_opts test...", logging.info)
 
-    numa_nodes = params.get("numa_nodes")
-    if numa_nodes:
-        numa_nodes = int(params.get("numa_nodes"))
-        if len(numa) != numa_nodes:
-            test.fail(
-                "Wrong number of numa nodes: %d. Expected: %d" %
-                (len(numa), numa_nodes))
+    # Get numa info from monitor
+    numa_monitor = vm.monitors[0].info_numa()
+    error_context.context("numa info in monitor: %r" % numa_monitor, logging.info)
 
-    for nodenr, node in enumerate(numa):
+    if os_type == 'linux':
+        # Get numa info in guest os, only for Linux
+        numa_guest = numa_info_guest()
+        error_context.context("numa info in guest: %r" % numa_guest, logging.info)
+
+        # Use 30 plus the gap of 'MemTotal' in OS and '-m' in cli as threshold
+        MemTotal = get_mem_info(session, 'MemTotal')
+        MemTotal = float(normalize_data_size('%s KB' % MemTotal))
+        error_context.context("MemTotal in guest os is %s MB"
+                              % MemTotal, logging.info)
+        threshold = float(params.get_numeric("mem") - MemTotal) + 30
+        error_context.context("The acceptable threshold is: %s"
+                              % threshold, logging.info)
+    else:
+        numa_guest = numa_monitor
+    session.close()
+
+    monitor_expect_nodes = params.get_numeric("monitor_expect_nodes")
+    if len(numa_monitor) != monitor_expect_nodes:
+        test.fail("[Monitor]Wrong number of numa nodes: %d. Expected: %d" %
+                  (len(numa_monitor), monitor_expect_nodes))
+
+    if os_type == 'linux':
+        guest_expect_nodes = int(params.get("guest_expect_nodes",
+                                            monitor_expect_nodes))
+        if len(numa_guest) != guest_expect_nodes:
+            test.fail("[Guest]Wrong number of numa nodes: %d. Expected: %d" %
+                      (len(numa_guest), guest_expect_nodes))
```

Comment:
L81~L85 move to L70?

Reply:
Updated, change the get info and compare info order to save one if block, please help to review again, thanks!

And update the cfg to indent 4 spaces.